migrate/migrate.c

```diff
@@ -0,0 +1,1073 @@
+      return -1;
+    }
+
+  return 0;
+}
+
+static int
+migrate_queries ()
+{
+  DB_QUERY_RESULT *result;
+  DB_VALUE value;
+  int i, error;
+
+  /* generated query */
+  error = migrate_generated (rename_query, 1);
+  if (error < 0)
+    {
+      fprintf (out, "migrate: execute query failed \"%s\"\n", rename_query);
+      return -1;
+    }
+
+  /* catalog query */
+  for (i = 0; i < sizeof (catalog_query) / sizeof (const char *); i++)
+    {
+      error = migrate_execute_query (catalog_query[i]);
+      if (error < 0)
+	{
+	  return -1;
+	}
+    }
+
+  error = migrate_execute_query (update_db_class_not_for_system_classes);
+  if (error < 0)
+    {
+      return -1;
+    }
+
+  error = migrate_generated (serial_query, 1);
+  if (error < 0)
+    {
+      fprintf (out, "migrate: execute query failed \"%s\"\n", serial_query);
+      return -1;
+    }
+
+  for (i = 0; i < sizeof (update_serial) / sizeof (const char *); i++)
+    {
+      error = migrate_execute_query (update_serial[i]);
+      if (error < 0)
+	{
+	  return -1;
+	}
+    }
+
+  error = migrate_execute_query (update_trigger);
+  if (error < 0)
+    {
+      return -1;
+    }
+
+  /* index query */
+  for (i = 0; i < sizeof (index_query) / sizeof (const char *); i++)
+    {
+      error = migrate_execute_query (index_query[i]);
+      if (error < 0)
+	{
+	  return -1;
+	}
+    }
+
+  error = migrate_execute_query (update_db_class_for_system_classes);
+  if (error < 0)
+    {
+      return -1;
+    }
+
+  return 0;
+}
+
+static void
+print_usage (void)
+{
+  printf ("usage: migrate [options] db-name\n\n");
+  printf ("valid options\n");
+  printf ("  -v:\t\tverbose detailed\n");
+  printf ("  -o FILE:\tredirect output to FILE\n");
+}
+
+static char *
+parse_args (int argc, char *argv[])
+{
+  int i;
+  char *dbname = NULL;
+  char *outfile = { 0 };
+
+  out = stdout;
+
+  for (i = 1; i < argc; i++)
+    {
+      if (argv[i][0] == '-')
+	{
+	  switch (argv[i][1])
+	    {
+	    case 'v':
+	      verbose = 1;
+	      break;
+	    case 'o':
+	      if (argc > i + 1)
+		{
+		  outfile = argv[++i];
+		  if (*outfile == '-')
+		    {
+		      printf ("error: missing output file\n\n");
+		      goto error_exit;
+		    }
+		  else
+		    {
+		      out = fopen (outfile, "w");
+		      if (out == NULL)
+			{
+			  printf ("error: can not open out file\n\n");
+			  goto error_exit;
+			}
+		    }
+		}
+	      else
+		{
+		  printf ("error: missing out file\n\n");
+		  goto error_exit;
+		}
+	      break;
+	    default:
+	      printf ("error: no such option\n\n");
+	      goto error_exit;
+	    }
+	}
+      else
+	{
+	  if (dbname == NULL)
+	    {
+	      dbname = argv[i];
+	    }
+	  else
+	    {
+	      printf ("error: not allowed multi DB\n\n");
+	      goto error_exit;
+	    }
+	}
+    }
+
+  if (dbname == NULL)
+    {
+      printf ("error: missing db-name\n\n");
+      goto error_exit;
+    }
+
+  return dbname;
+
+error_exit:
+  print_usage ();
+  return NULL;
```

Comment:
You need to close outfile if error as the following.
if (out != NULL && out != stdout)
  {
      fclose(out);
      out = stdout;
  }